doma-core/src/main/java/org/seasar/doma/jdbc/dialect/PostgreSqlUpsertAssembler.java

```diff
@@ -0,0 +1,97 @@
+package org.seasar.doma.jdbc.dialect;
+
+import java.util.List;
+import org.seasar.doma.internal.jdbc.sql.PreparedSqlBuilder;
+import org.seasar.doma.jdbc.InParameter;
+import org.seasar.doma.jdbc.criteria.tuple.Tuple2;
+import org.seasar.doma.jdbc.entity.EntityPropertyType;
+import org.seasar.doma.jdbc.entity.EntityType;
+import org.seasar.doma.jdbc.query.DuplicateKeyType;
+import org.seasar.doma.jdbc.query.UpsertAssembler;
+import org.seasar.doma.jdbc.query.UpsertAssemblerContext;
+import org.seasar.doma.jdbc.query.UpsertAssemblerSupport;
+import org.seasar.doma.jdbc.query.UpsertSetValue;
+
+public class PostgreSqlUpsertAssembler implements UpsertAssembler {
+  private final PreparedSqlBuilder buf;
+  private final EntityType<?> entityType;
+  private final DuplicateKeyType duplicateKeyType;
+  private final UpsertAssemblerSupport upsertAssemblerSupport;
+  private final List<EntityPropertyType<?, ?>> keys;
+  private final List<Tuple2<EntityPropertyType<?, ?>, InParameter<?>>> insertValues;
+  private final List<Tuple2<EntityPropertyType<?, ?>, UpsertSetValue>> setValues;
+
+  public PostgreSqlUpsertAssembler(UpsertAssemblerContext context) {
+    this.buf = context.buf;
+    this.entityType = context.entityType;
+    this.duplicateKeyType = context.duplicateKeyType;
+    this.keys = context.keys;
+    this.insertValues = context.insertValues;
+    this.setValues = context.setValues;
+    this.upsertAssemblerSupport = new UpsertAssemblerSupport(context.naming, context.dialect);
+  }
+
+  @Override
+  public void assemble() {
+    buf.appendSql("insert into ");
+    tableNameAndAlias(entityType);
+    buf.appendSql(" (");
+    for (Tuple2<EntityPropertyType<?, ?>, InParameter<?>> insertValue : insertValues) {
+      column(insertValue.component1());
+      buf.appendSql(", ");
+    }
+    buf.cutBackSql(2);
+    buf.appendSql(") values (");
+    for (Tuple2<EntityPropertyType<?, ?>, InParameter<?>> insertValue : insertValues) {
+      buf.appendParameter(insertValue.component2());
+      buf.appendSql(", ");
+    }
+    buf.cutBackSql(2);
+    buf.appendSql(") on conflict (");
+    for (EntityPropertyType<?, ?> key : keys) {
+      column(key);
+      buf.appendSql(", ");
+    }
+    buf.cutBackSql(2);
+    buf.appendSql(")");
+    if (duplicateKeyType == DuplicateKeyType.IGNORE) {
+      buf.appendSql(" do nothing");
+    } else if (duplicateKeyType == DuplicateKeyType.UPDATE) {
+      buf.appendSql(" do update set ");
+      for (Tuple2<EntityPropertyType<?, ?>, UpsertSetValue> setValue : setValues) {
+        column(setValue.component1());
+        buf.appendSql(" = ");
+        setValue.component2().accept(new UpsertSetValueVisitor());
```

Comment:
It is better to move the instantiation of `UpsertSetValueVisitor` outside of the loop.